### PR TITLE
Refactor WS integration to use common publisher

### PR DESCRIPTION
**Pull request by ssd04**

### CODE REVIEW

File: disabled/disabledHub.go [hunk 1]
```diff
@@ -9,33 +9,33 @@ import (
...
-// Run does nothing
-func (h *Hub) Run() error {
+// RegisterListener does nothing
+func (h *Hub) RegisterListener() error {
 	return nil
 }
 
-// Broadcast does nothing
-func (h *Hub) Broadcast(_ data.BlockEvents) {
+// Publish -
```
Comment (sstanculeanu) [inline disabled/disabledHub.go:17]: keep the `does nothing` comments?

File: dispatcher/hub/commonHub.go [hunk 2]
```diff
@@ -308,7 +235,7 @@ func (ch *commonHub) handleBlockEventsWithOrderBroadcast(blockTxs data.BlockEven
 	}
 }
 
-func (ch *commonHub) handleScrsBroadcast(blockScrs data.BlockScrs) {
+func (ch *commonHub) PublishScrs(blockScrs data.BlockScrs) {
```
Comment (sstanculeanu) [inline dispatcher/hub/commonHub.go:238]: missing comment

File: factory/pubsubFactory.go [hunk 3]
```diff
@@ -47,10 +50,11 @@ func createRabbitMqPublisher(config config.RabbitMQConfig, marshaller marshal.Ma
...
-		Filter:             filters.NewDefaultFilter(),
-		SubscriptionMapper: dispatcher.NewSubscriptionMapper(),
+func createWSPublisher(commonHub dispatcher.Hub) (process.Publisher, error) {
+	publisher, err := process.NewPublisher(commonHub)
+	if err != nil {
+		return nil, err
 	}
-	return hub.NewCommonHub(args)
+
+	return publisher, nil
```
Comment (sstanculeanu) [inline factory/pubsubFactory.go:59]: return process.NewPublisher(commonHub)

File: dispatcher/hub/commonHub.go [hunk 4]
```diff
@@ -73,45 +63,30 @@ func checkArgs(args ArgsCommonHub) error {
 	return nil
 }
 
-// Run is launched as a goroutine and listens for events on the exposed channels
-// TODO: use protection from triggering multiple times
-func (ch *commonHub) Run() error {
+// RegisterListener creates a goroutine and listens for WS register events
+func (ch *commonHub) RegisterListener() error {
```
Comment (sstanculeanu) [inline dispatcher/hub/commonHub.go:67]: why not launch the go routine on constructor?

File: dispatcher/hub/commonHub.go (hunk 2 above)
Comment (ssd04) [reply dispatcher/hub/commonHub.go:238]: added comment

File: factory/pubsubFactory.go (hunk 3 above)
Comment (ssd04) [reply factory/pubsubFactory.go:59]: done

File: dispatcher/hub/commonHub.go (hunk 4 above)
Comment (ssd04) [reply dispatcher/hub/commonHub.go:67]: this component is using in 2 ways, firstly to handle websocket subscriptions, and secondly as a handler for pushing block events (which runs in the common publisher goroutine); so I wanted to trigger it separately to be very clear, as for the others in notifierRunner

**Review by sstanculeanu — APPROVED** [review]

File: tools/testData.go [hunk 5]
```diff
@@ -1,17 +1,30 @@
...
+	"github.com/multiversx/mx-chain-core-go/marshal"
+	"github.com/multiversx/mx-chain-notifier-go/common"
 	notifierData "github.com/multiversx/mx-chain-notifier-go/data"
 )
 
+type blockData struct {
+	marshaller marshal.Marshalizer
+}
+
+func NewBlockData(marshaller marshal.Marshalizer) (*blockData, error) {
```
Comment (sstanculeanu) [inline tools/testData.go:18]: missing comment

File: dispatcher/hub/commonHub.go [hunk 6]
```diff
@@ -73,45 +63,30 @@ func checkArgs(args ArgsCommonHub) error {
...
-			ch.handleTxsBroadcast(txsEvent)
-
-		case txsEvent := <-ch.broadcastBlockEventsWithOrder:
-			ch.handleBlockEventsWithOrderBroadcast(txsEvent)
-
-		case scrsEvent := <-ch.broadcastScrs:
-			ch.handleScrsBroadcast(scrsEvent)
-
 		case dispatcherClient := <-ch.register:
 			ch.registerDispatcher(dispatcherClient)
```
Comment (iulianpascalau) [inline dispatcher/hub/commonHub.go:89]: why bother using a go-routine ran process loop just to handle the register/unregister event and we also use mutex operations on the ch.registerDispatcher and ch.unregisterDispatcher functions? The go routines & the 2 channels can be safely removed. The function RegisterEvent can call ch.registerDispatcher  and UnregisterEvent can call directly ch.unregisterDispatcher

File: tools/testData.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ssd04) [reply tools/testData.go:18]: added missing comment

File: dispatcher/hub/commonHub.go (hunk 6 above)
Comment (ssd04) [reply dispatcher/hub/commonHub.go:89]: right, removed the loop

File: dispatcher/hub/commonHub.go [hunk 7]
```diff
@@ -358,6 +286,9 @@ func (ch *commonHub) unregisterDispatcher(d dispatcher.EventDispatcher) {
 
 // Close will close the goroutine and channels
 func (ch *commonHub) Close() error {
+	ch.mutState.Lock()
```
[the remark this answers is not in the document]
Comment (ssd04) [reply dispatcher/hub/commonHub.go:289]: removed this part also after removing listener loop

File: dispatcher/hub/commonHub_test.go [hunk 8]
```diff
@@ -66,7 +98,7 @@ func TestCommonHub_RegisterDispatcher(t *testing.T) {
 	dispatcher1 := mocks.NewDispatcherMock(nil, hub)
 	dispatcher2 := mocks.NewDispatcherMock(nil, hub)
 
-	hub.Run()
+	err = hub.RegisterListener()
```
[the remark this answers is not in the document]
Comment (ssd04) [reply dispatcher/hub/commonHub_test.go:101]: removed this call, so ne need for update anymore

File: dispatcher/hub/commonHub.go [hunk 9]
```diff
@@ -192,7 +114,8 @@ func (ch *commonHub) UnregisterEvent(event dispatcher.EventDispatcher) {
 	}
 }
 
-func (ch *commonHub) handleBroadcast(blockEvents data.BlockEvents) {
+// Publish will publish logs and events to dispatcher
+func (ch *commonHub) Publish(blockEvents data.BlockEvents) {
 	subscriptions := ch.subscriptionMapper.Subscriptions()
 
 	for _, subscription := range subscriptions {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply dispatcher/hub/commonHub.go:81]: it cannot be set to return just a map, since there can be multiple subscription events per type

**Review by bogdan-rosianu — DISMISSED** [review]

File: dispatcher/hub/commonHub.go [hunk 10]
```diff
@@ -22,20 +21,14 @@ type ArgsCommonHub struct {
...
-	broadcastTxs                  chan data.BlockTxs
-	broadcastBlockEventsWithOrder chan data.BlockEventsWithOrder
-	broadcastScrs                 chan data.BlockScrs
-	closeChan                     chan struct{}
-	cancelFunc                    func()
+	filter             filters.EventFilter
+	subscriptionMapper dispatcher.SubscriptionMapperHandler
+	mutDispatchers     sync.RWMutex
+	dispatchers        map[uuid.UUID]dispatcher.EventDispatcher
+	register           chan dispatcher.EventDispatcher
```
Comment (iulianpascalau) [inline dispatcher/hub/commonHub.go:28]: unused `register` and `unregister` channels?

File: dispatcher/hub/commonHub.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply dispatcher/hub/commonHub.go:81]: can still be refactored and the filtering removed

File: dispatcher/hub/commonHub.go (hunk 10 above)
Comment (ssd04) [reply dispatcher/hub/commonHub.go]: removed unused code

File: dispatcher/hub/commonHub.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (ssd04) [reply dispatcher/hub/commonHub.go]: refactored subscription mapper

**Review by iulianpascalau — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Pull request merged**